content/projects/gsoc/2020/projects/external-fingerprint-storage.adoc

```diff
@@ -15,81 +15,263 @@ links:
+
+* `void save()`
+
+** Saves the given Fingerprint in the storage.
+
+* `Fingerprint load(String id)`
+
+** Returns the Fingerprint with the given unique ID. The unique ID for a fingerprint is defined by
+`Fingerprint#getHashString()`.
+
+* `void delete(String id)`
+
+** Deletes the Fingerprint with the given unique ID.
+
+* `boolean isReady()`
+
+** Returns true if there is some data in the fingerprint database corresponding to the particular Jenkins instance.
+
+For more details, please refer to the Javadoc:
+
+* link:https://javadoc.jenkins.io/jenkins/fingerprints/FingerprintStorage.html[FingerprintStorage]
+
+* link:https://javadoc.jenkins.io/jenkins/fingerprints/FingerprintStorageDescriptor.html[FingerprintStorageDescriptor]
+
+==== Fingerprint Cleanup
+
+Fingerprint cleanup thread works by periodically iterating over the fingerprints and editing the job and build
+information of the ones based on whether they are still present in the system.
+It also deletes the fingerprints which do not have any build or job associated with them.
+
+We extend this fingerprint cleanup functionality to be supported by external storages.
+Fingerprint cleanup support for external storage plugins was implemented in
+link:https://www.jenkins.io/changelog/#v2.242[Jenkins-2.248].
+FingerprintStorage API was extended with the following methods:
+
+* `iterateAndCleanupFingerprints(TaskListener taskListener)`
+
+** Plugins can implement this method (which is called by Jenkins core periodically) to iterate and cleanup the
+fingerprints.
+The reason to design it this way, and not to iterate all the fingerprints via core, is because external storages
+may be able to implement more efficient traversal strategies on their own.
+
+* `boolean cleanFingerprint(Fingerprint fingerprint, TaskListener taskListener)`
+
+** This provides a reference implementation of cleanup, which external storages can use to cleanup a fingerprint.
+They may use this, or extend it to provide custom implementations.
+
+This allows the plugins to implement their own cleanup strategies in efficient ways.
+For example, the link:https://github.com/jenkinsci/redis-fingerprint-storage-plugin[Redis plugin] uses
+link:https://redis.io/commands/scan[cursors] to traverse and cleanup the fingerprints.
+
+Finally, we introduced the option to turn off fingerprint cleanup.
+This was done because it may be the case that storing extra data may be cheaper than performing cleanups,
+especially with external storages.
+
+==== Fingerprint Migration
+
+We implemented a lazy migration strategy to transfer the fingerprints from local storage to the newly
+configured external storage.
+Once an external fingerprint storage is configured, the new fingerprints are stored directly in the new storage
+engine.
+However, the old fingerprints present on the disk storage are migrated as and when they are used.
 
-=== Deliverables
```

Comment:
I would recommend to keep this section before "Getting the code". It would help readers to understand what was done on the top level before diving into details.